finalfusion-ffi/src/lib.rs

```diff
@@ -65,3 +71,42 @@ pub unsafe extern "C" fn ff_free_embeddings(embeddings: *mut Embeddings<VocabWra
         Box::from_raw(embeddings);
     }
 }
+
+/// Return the embedding dimensionality.
+#[no_mangle]
+pub unsafe extern "C" fn dims(embeddings: *const Embeddings<VocabWrap, StorageWrap>) -> usize {
+    check_null!(embeddings);
+    let embeddings = &*embeddings;
+    embeddings.storage().shape().1
+}
+
+/// Embedding Lookup.
+///
+/// Writes the embedding to `buffer`. Callers need to verify that `buffer` is large enough for the
+/// embedding dimensionality.
+///
+/// Returns an integer indicating whether an embedding could be found and anything was written to
+/// `buffer`.
+///
+/// **Note:** Passing a buffer with incorrect size will lead to UB. The correct dimensionality can
+/// be obtained through `dims()`.
+#[no_mangle]
+pub unsafe extern "C" fn embedding(
+    embeddings: *const Embeddings<VocabWrap, StorageWrap>,
+    word: *const c_char,
+    buffer: *mut f32,
```

Comment:
This overwrites memory if `buffer` is too small. Of course, the caller must get the correct size with the dimensions function. It is worth pondering whether we'd want to return the embedding as an array instead. Benefits: always has the correct size. Downside: requires an extra allocation, whereas here the caller can always the buffer. Another possibility would be to `realloc` buffer to have the right size, but I guess it's not nice to change the size behind the caller's back.

Rename `buffer` to `embedding`.

Reply:
(I am fine with the current signature though, just wanted to point out the trade-off.)

Reply:
I had written some comments about taking `*mut f32` or returning some `Embeddings` struct:
https://github.com/sfb833-a3/finalfusion_tensorflow/pull/6#issuecomment-524922748

In a nutshell: 
 - using the buffer puts the burden on the caller to allocate `embeddings` with at least `dims`  space
- returning the pointer to the buffer of a `Vec` requires calling back into Rust to properly drop the `Vec`
- in a naive implementation as in return *mut float, the caller needs to a) pass `Embeddings` to get the right dimension for deallocation (note, we need to trim the buffer before returning it, else we don't know the capacity)
- different solution is defining an `Embedding` struct holding `len`, (potentially `capacity`) and `*mut float` for which we define a `free` method in Rust. Here we don't need to rely on having valid `Embeddings` to deallocate the `Embedding` but we're now returning a structure into C.

I thought writing to the buffer was the most elegant and flexible solution of the above. In any way we go about this, the caller can mess up. The trade-off would be possible memory leaks or overwriting memory, not sure which one's preferable.

Reply:
>     * returning the pointer to the buffer of a `Vec` requires calling back into Rust to properly drop the `Vec`

Not needed, you can `mem::forget` the `Vec` after getting the pointer. If the `Vec` is on the stack, the capacity and length will be deallocated when leaving the stack frame. The caller can free the memory after use.
 
>     * in a naive implementation as in return *mut float, the caller needs to a) pass `Embeddings` to get the right dimension for deallocation 

Nope. `free` does not require the length, only a pointer, and it will deallocate the memory. The only potential problem is that the memory could be allocated with a different allocator, but this should not be a problem anymore since Rust switched from jealloc to the the system allocator.

> I thought writing to the buffer was the most elegant and flexible solution of the above. In any way we go about this, the caller can mess up. The trade-off would be possible memory leaks or overwriting memory, not sure which one's preferable.

Memory leaks are always preferable over overwriting memory, since that is undefined behavior.

Reply:
> Nope. `free` does not require the length, only a pointer. The only potential problem is that the memory could be allocated with a different allocator, but this should not be a problem anymore since Rust switched from jealloc to the the system allocator.

Oh, nice. I was thinking about `Vec::from_raw` which requires both `len` and `capacity`.

So, which way do we want to go? I doubt that the additional heap allocation + memcpy will be a bottleneck. Do you prefer returning an array?

Reply:
We are doing the allocation anyway in when a word is unknown, so I guess the penalty is not spectacular. So let's do it ;).

Reply:
I `perf top`'d a loop looking up ~50k words in tensorflow on the weekend and most time was spent in `add_assign` followed by a bunch of Python calls. So I doubt the allocation plays a role. I changed that method to return `*mut float`.